bpf/lib/nat.h

```diff
@@ -552,6 +554,60 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
 			tuple.sport = 0;
 			icmp_echoreply = true;
 			break;
+		case ICMP_DEST_UNREACH:
+			switch (icmphdr.code) {
+			case ICMP_FRAG_NEEDED:
+				/* According to the RFC 5508, any networking
+				 * equipment that is responding with an ICMP Error
+				 * packet should embed the original packet in its
+				 * response.
+				 */
+
+				off += sizeof(icmphdr);
+				if (ctx_load_bytes(ctx, off, &iphdr,
+						   sizeof(iphdr)) < 0)
```

Comment:
On XDP, these `ctx_load_bytes` are not converted into a helper call [but using `memcpy`](https://github.com/cilium/cilium/blob/82c742f2e9fb65fb3fc392c7dcb3b4e22b69c650/bpf/include/bpf/ctx/xdp.h#L34). That can be quite expensive in the number of instructions used. Given we only need four fields from the IP header, maybe we could simply read those four fields? (See for instance how we load the L4 fields we need below.)

Reply:
@sahid Answering here, so we can close the thread once resolved. 

> The saddr and daddr are at the very end, we will still have to load the full object, did I missed something?

If you only want to load e.g. the `saddr` and `daddr`, you can change `off` to start reading for the offset where `saddr` is and change `sizeof(iphdr)` to read only 8 bytes (2 IPv4 addresses). No need to read `sizeof(iphdr)` bytes.

Reply:
Oh right! I think I see where you want me to go now. Thank!

Reply:
I have made change for IPv4 only, could you let me know your thinking about it when you have a moment?

Reply:
Looks good but I wouldn't put it all inside one if condition and we also shouldn't hardcode the sizes.

Reply:
ACK I will fix that and work on IPv6. Thanks.

Reply:
I fixed you point and addressed IPv6. 

BTW and I should probably have asked this question before, since we are now calling `ctx_load_bytes` more times, we are also going to call `memcpy` more times, that is not going to increase the number of instructions?

Reply:
Yep, I'd expect the number of instructions for the three `memcpy`s to be less than the number of instructions for the single larger `memcpy`. Note however that, on the tc side, `ctx_load_bytes` is converted into helper calls, so this change will lead to more instructions for tc programs. Hopefully, we're not just trading a program size issue on XDP for one on tc :crossed_fingers:

Reply:
ACK thank you Paul.

Reply:
Don't you think that I should revert back the change as it was before?